### PR TITLE
[WIP][runtime] Use mkbundle to test statically-linked AOT

### PR DESCRIPTION
**Pull request by alexanderkyte**
Supporting infrastructure to test dedup changes

### CODE REVIEW

File: mcs/build/rules.make [hunk 1]
```diff
@@ -193,6 +193,32 @@ endif
 
 endif #ifneq ("$(wildcard $(topdir)/class/lib/$(PROFILE))","")
 
+MKBUNDLE_TEST_BIN = $(TEST_HARNESS).static
```
Comment (marek-safar) [inline mcs/build/rules.make]: Tests specific code needs to go to tests.make

File: mcs/build/rules.make [hunk 2]
```diff
@@ -193,6 +193,32 @@ endif
...
+BUNDLED_ASSEMBLIES := $(sort $(patsubst .//%,%,$(filter-out %.exe.static %.dll.dll %.exe.dll %bare% %plaincore% %secxml% %Facades% %ilasm%,$(filter %.dll %.exe,$(wildcard $(topdir)/class/lib/$(PROFILE)/*)))))
+
+$(MKBUNDLE_EXE): $(topdir)/tools/mkbundle/mkbundle.cs
+	make -C $(topdir)/tools/mkbundle
+
+mkbundle-all-tests: $(MKBUNDLE_EXE)
+	$(Q_AOT) $(MAKE) do-test
+	$(Q_AOT) $(MAKE) $(MKBUNDLE_TEST_BIN) # recursive make re-computes variables for BUNDLED_ASSEMBLIES
+
+ifdef MKBUNDLE_DEDUP
```
Comment (marek-safar) [inline mcs/build/rules.make]: What this is for? It does not appaer to be used

File: mcs/class/System.IdentityModel/testing_aot_full_System.IdentityModel_test.dll.sources [hunk 3]
```diff
@@ -0,0 +1 @@
+# Skipped
```
Comment (marek-safar) [inline mcs/class/System.IdentityModel/testing_aot_full_System.IdentityModel_test.dll.sources]: It'd be cleaner to use `NO_TEST` in Makefile

File: mcs/tools/mkbundle/mkbundle.cs [hunk 4]
```diff
@@ -868,6 +937,33 @@ static void GenerateBundles (List<string> files)
...
+				tc.WriteLine ("\textern const void *mono_aot_module_{0}_info;", asm);
+			}
+
+			tc.WriteLine ("\nstatic void install_aot_modules (void) {\n");
+			foreach (string asm in aot_names){
+				tc.WriteLine ("\tmono_aot_register_module (mono_aot_module_{0}_info);\n", asm);
+			}
+
+			string enum_aot_mode;
+			if (aot_mode == "full") {
```
Comment (marek-safar) [inline mcs/tools/mkbundle/mkbundle.cs]: `switch` is preferred

File: mcs/tools/mkbundle/mkbundle.cs [hunk 5]
```diff
@@ -1034,10 +1078,17 @@ static void GenerateBundles (List<string> files)
 						smonolib = "`pkg-config --variable=libdir mono-2`/libmono-2.0.a ";
 					else
 						smonolib = "-Wl,-Bstatic -lmono-2.0 -Wl,-Bdynamic ";
+
+					var aot_libs = new StringBuilder ("");
+					foreach (var apath in aot_paths) {
+						aot_libs.Append (apath);
+						aot_libs.Append (" ");
+					}
```
Comment (akoeplinger) [inline mcs/tools/mkbundle/mkbundle.cs]: this could just be `var aot_libs = String.Join (" ", aot_paths);`

File: mcs/tools/mkbundle/mkbundle.cs [hunk 6]
```diff
@@ -1411,6 +1414,15 @@ static void AotCompile (List<string> files)
 		if (aot_runtime == null)
 			aot_runtime = runtime;
 
+		if (aot_runtime == null || aot_runtime.Length == 0){
+			if (IsUnix)
+				aot_runtime = Process.GetCurrentProcess().MainModule.FileName;
```
Comment (akoeplinger) [inline mcs/tools/mkbundle/mkbundle.cs]: please use curly braces here to match the else.

File: mcs/class/System.Security/testing_aot_full_System.Security_test.dll.sources [hunk 7]
```diff
@@ -0,0 +1 @@
+# Skipped
```
Comment (akoeplinger) [inline mcs/class/System.Security/testing_aot_full_System.Security_test.dll.sources]: hmm, why are they not executed? do they not work on fullaot?

File: mcs/build/tests.make [hunk 8]
```diff
@@ -153,8 +153,8 @@ BUNDLED_ASSEMBLIES := $(sort $(patsubst .//%,%,$(filter-out %.exe.static %.dll.d
 $(MKBUNDLE_EXE):
 	make -C $(topdir)/tools/mkbundle
 
-$(TEST_HARNESS_BIN): $(MKBUNDLE_EXE)
-	$(TEST_RUNTIME) $(MKBUNDLE_EXE) -L $(topdir)/class/lib/$(PROFILE) -v --deps $(TEST_HARNESS) $(BUNDLED_ASSEMBLIES) -o $(TEST_HARNESS_BIN) --aot-mode $(STATIC_AOT_BUNDLE) --aot-runtime $(RUNTIME)
+$(TEST_HARNESS_BIN): $(MKBUNDLE_EXE) $(test_assemblies) $(BUNDLED_ASSEMBLIES)
+	PKG_CONFIG_PATH="$(topdir)/../data" $(TEST_RUNTIME) $(MKBUNDLE_EXE) -L $(topdir)/class/lib/$(PROFILE) -v --deps $(TEST_HARNESS) $(test_assemblies) $(BUNDLED_ASSEMBLIES) -o $(TEST_HARNESS_BIN) --aot-mode $(AOT_MODE) --aot-runtime $(RUNTIME) --aot-args $(AOT_BUILD_ATTRS) --in-tree $(topdir)/.. --keeptemp
```
Comment (akoeplinger) [inline mcs/build/tests.make]: which pkg-config files do you need?

I think we should use `$(topdir)/../runtime/_tmpinst` instead, we can add the necessary pkg-config files there, check the runtime/Makefile.am file.

File: mcs/class/System.IdentityModel/testing_aot_full_System.IdentityModel_test.dll.sources (hunk 3 above)
Comment (akoeplinger) [inline mcs/class/System.IdentityModel/testing_aot_full_System.IdentityModel_test.dll.sources]: 👍 , though I have the same question as for System.Security: why don't we run the tests on fullaot?

File: mcs/build/tests.make [hunk 9]
```diff
@@ -96,8 +96,10 @@ endif
 
 test_assemblies :=
 
+test_lib_output = $(topdir)/class/lib/$(PROFILE)/$(test_lib)
```
Comment (akoeplinger) [inline mcs/build/tests.make]: this won't work, a lot of the tests look up resources and other test stuff relative to the test assembly location and won't find those files anymore.

File: mcs/class/System.IdentityModel/testing_aot_full_System.IdentityModel_test.dll.sources (hunk 3 above)
Comment (alexanderkyte) [inline mcs/class/System.IdentityModel/testing_aot_full_System.IdentityModel_test.dll.sources]: This is the current behavior, we exclude it from the list of things to test on the nunitlite XI test harness that was the basis for fullaot's test subset. That list was the basis for the one in scripts/ci so I never made the .excludes/.sources file for it before. I can whitelist some tests and add them back if you'd like to add System.Security as a testing target.

File: mcs/build/rules.make (hunk 2 above)
Comment (alexanderkyte) [inline mcs/build/rules.make]: Not yet it isn't. I'll split that into the other PR.

File: mcs/build/tests.make (hunk 9 above)
Comment (alexanderkyte) [inline mcs/build/tests.make]: If the tests are not currently relocatable, this PR is going to need to change that as well if we hope to run them statically linked

Comment (alexanderkyte) [inline mcs/build/tests.make]: How does https://github.com/xamarin/Touch.Unit handle that?

**Comment by vargaz** [comment]
This references the aot dedup functionality which is not yet committed.

**Comment by alexanderkyte** [comment]
Yeah, this is WIP

File: mcs/build/tests.make (hunk 9 above)
Comment (alexanderkyte) [inline mcs/build/tests.make]: After changing the resource includes in the Makefiles, I didn't run into any tests hit by the jenkins CI script that had any issues with the movement on the fullaot profile. The resources are bundled right into the dll. We'll have to see if any failures on other profiles pop up. The relative paths can be fixed. The tests are being run from the same directories, for what it's worth.

File: mcs/class/System.XML/Test/System.Xml/XmlUrlResolverTests.cs [hunk 10]
```diff
@@ -97,6 +97,7 @@ public void NullAbsoluteUriWithCustomSchemedRelativeUri ()
 		}
 
 		[Test]
+		[Category ("StaticNotWorking")] // Can't find .dll files when bundled in .exe
```
Comment (marek-safar) [inline mcs/class/System.XML/Test/System.Xml/XmlUrlResolverTests.cs]: What is static, please use unambiguous names

Comment (alexanderkyte) [inline mcs/class/System.XML/Test/System.Xml/XmlUrlResolverTests.cs]: The assemblies are being semantically statically linked together, and the AOTed assemblies are being statically linked, and the runtine is being statically linked into that. So everything is static. I can make it "StaticLinkNotSupported" or "StaticAotNotSupported" if either of those seem better to you. It's unfortunate that the dynamically linked​ "mobile static" profile was named such.

Comment (marek-safar) [inline mcs/class/System.XML/Test/System.Xml/XmlUrlResolverTests.cs]: `StaticLinkedAotNotWorking` seems reasonable as for example this test case can be make to work

**Pull request closed without merging**